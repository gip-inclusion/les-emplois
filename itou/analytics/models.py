import uuid

from django.db import models
from django.utils import timezone

from itou.common_apps.address.departments import DEPARTMENTS
from itou.users.enums import UserKind


class DatumCode(models.TextChoices):
    # Employee record - Base
    EMPLOYEE_RECORD_COUNT = "ER-001", "FS totales"
    EMPLOYEE_RECORD_DELETED = "ER-002", "FS (probablement) supprimées"
    # Employee record - Lifecycle
    EMPLOYEE_RECORD_PROCESSED_AT_FIRST_EXCHANGE = "ER-101", "FS intégrées (0000) au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_AT_FIRST_EXCHANGE = "ER-102", "FS avec une erreur au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_3436_AT_FIRST_EXCHANGE = "ER-102-3436", "FS avec une erreur 3436 au premier retour"
    EMPLOYEE_RECORD_WITH_AT_LEAST_ONE_ERROR = "ER-103", "FS ayant eu au moins un retour en erreur"
    # Approval - Base
    APPROVAL_COUNT = "AP-001", "PASS IAE total"
    APPROVAL_CANCELLED = "AP-002", "PASS IAE annulés"
    # Approval - PE notification cycle
    APPROVAL_PE_NOTIFY_SUCCESS = "AP-101", "PASS IAE synchronisés avec succès avec pole emploi"
    APPROVAL_PE_NOTIFY_PENDING = "AP-102", "PASS IAE en attente de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_ERROR = "AP-103", "PASS IAE en erreur de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_READY = "AP-104", "PASS IAE prêts à être synchronisés avec pole emploi"
    # Users
    USER_COUNT = "US-001", "Nombre d'utilisateurs"
    USER_JOB_SEEKER_COUNT = "US-011", "Nombre de demandeurs d'emploi"
    USER_PRESCRIBER_COUNT = "US-012", "Nombre de prescripteurs"
    USER_EMPLOYER_COUNT = "US-013", "Nombre d'employeurs"
    USER_LABOR_INSPECTOR_COUNT = "US-014", "Nombre d'inspecteurs du travail"
    USER_ITOU_STAFF_COUNT = "US-015", "Nombre d'administrateurs"
    # API usage
    API_TOTAL_CALLS = "API-001", "API : total d'appels reçus"
    API_TOTAL_UV = "API-002", "API : total de visiteurs uniques"
    API_TOTAL_CALLS_CANDIDATS = "API-003", "API candidats : total d'appels reçus"
    API_TOTAL_UV_CANDIDATS = "API-004", "API candidats : total de visiteurs uniques"
    API_TOTAL_CALLS_GEIQ = "API-005", "API GEIQ : total d'appels reçus"
    API_TOTAL_UV_GEIQ = "API-006", "API GEIQ : total de visiteurs uniques"
    API_TOTAL_CALLS_ER = "API-007", "API FS : total d'appels reçus"
    API_TOTAL_UV_ER = "API-008", "API FS : total de visiteurs uniques"
    API_TOTAL_CALLS_MARCHE = "API-009", "API Le marché : total d'appels reçus"
    API_TOTAL_CALLS_SIAES = "API-010", "API siaes : total d'appels reçus"
    API_TOTAL_UV_SIAES = "API-011", "API siaes : total de visiteurs uniques (par adresse IP)"
    API_TOTAL_CALLS_STRUCTURES = "API-012", "API structures : total d'appels reçus"


class Datum(models.Model):
    """Store an aggregated `value` of the `code` data point for the specified `bucket`."""

    id = models.UUIDField(primary_key=True, default=uuid.uuid4)

    code = models.TextField(choices=DatumCode.choices)
    bucket = models.TextField()
    value = models.IntegerField()  # Integer offers the best balance between range, storage size, and performance

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "data"
        unique_together = ["code", "bucket"]
        indexes = [models.Index(fields=["measured_at", "code"])]


class StatsDashboardVisit(models.Model):
    dashboard_id = models.IntegerField(verbose_name="ID tableau de bord Metabase")
    dashboard_name = models.TextField(verbose_name="nom de la vue du tableau de bord")
    department = models.CharField(verbose_name="département", choices=DEPARTMENTS.items(), max_length=3, null=True)
    region = models.TextField(verbose_name="région", null=True)
    current_company_id = models.IntegerField(verbose_name="ID entreprise courante", null=True)
    current_prescriber_organization_id = models.IntegerField(
        verbose_name="ID organisation prescriptrice courante", null=True
    )
    current_institution_id = models.IntegerField(verbose_name="ID institution courante", null=True)
    user_kind = models.TextField(verbose_name="type d'utilisateur", choices=UserKind.choices)
    user_id = models.IntegerField(verbose_name="ID utilisateur")

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "visite de tableau de bord"
