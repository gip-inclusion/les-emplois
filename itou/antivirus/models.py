from django.db import models
from django.utils.text import capfirst

from itou.files.models import File


class Scan(models.Model):
    file = models.OneToOneField(File, on_delete=models.CASCADE)
    clamav_signature = models.TextField()
    clamav_completed_at = models.DateTimeField(null=True, verbose_name="analyse ClamAV le")
    clamav_infected = models.BooleanField(null=True, verbose_name="fichier infect√© selon ClamAV")
    comment = models.TextField(blank=True, verbose_name="commentaire")

    class Meta:
        verbose_name = "analyse antivirus"
        verbose_name_plural = "analyses antivirus"

    def __str__(self):
        text = f"{capfirst(self._meta.verbose_name)} {self.file_id}"
        if self.clamav_infected:
            text = f"[VIRUS] {text}"
        return text
