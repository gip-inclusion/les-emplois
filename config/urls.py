from django.conf import settings
from django.contrib import admin
from django.urls import include, path, re_path, register_converter
from django.views.generic import TemplateView

from itou.utils.urls import SiretConverter
from itou.www.dashboard import views as dashboard_views
from itou.www.login import views as login_views
from itou.www.signup import views as signup_views


register_converter(SiretConverter, "siret")

urlpatterns = [
    path("admin/", admin.site.urls),
    # --------------------------------------------------------------------------------------
    # allauth URLs. Order is important because some URLs are overriden.
    # --------------------------------------------------------------------------------------
    # Override allauth `account_signup` URL.
    # /accounts/signup/ <=> account_signup
    # We don't want any user to be able to signup using the default allauth `signup` url
    # because we have multiple specific signup processes for different kind of users.
    re_path(r"^accounts/signup/$", signup_views.signup),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_login` URL.
    # /accounts/login/ <=> account_login
    # We override this view because the login page should look slightly differently
    # for job seekers, prescribers and employers.
    # Also, PEAMU is only available for job seekers.
    re_path(r"^accounts/login/$", login_views.login),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_change_password` URL.
    # /accounts/password/change/ <=> account_change_password
    # https://github.com/pennersr/django-allauth/issues/468
    re_path(r"^accounts/password/change/$", dashboard_views.password_change),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_reset_password` URL.
    # Avoid user enumeration via password reset page.
    re_path(r"^accounts/password/reset/$", signup_views.ItouPasswordResetView.as_view()),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_logout` URL.
    # /accounts/logout/ <=> account_logout
    # We need custom code to process PEAMU logout.
    re_path(r"^accounts/logout/$", dashboard_views.logout),
    # --------------------------------------------------------------------------------------
    # Other allauth URLs.
    path("accounts/", include("allauth.urls")),
    # --------------------------------------------------------------------------------------
    # PEAMU URLs.
    path("accounts/", include("itou.allauth_adapters.peamu.urls")),
    # FranceConnect URLs.
    path("franceconnect/", include("itou.france_connect.urls")),
    # --------------------------------------------------------------------------------------
    # API.
    path("api/v1/", include("itou.api.urls", namespace="v1")),
    path("api/", include("itou.www.api.urls")),
    # www.
    path("", include("itou.www.home.urls")),
    path("apply/", include("itou.www.apply.urls")),
    path("approvals/", include("itou.www.approvals_views.urls")),
    path("autocomplete/", include("itou.www.autocomplete.urls")),
    path("dashboard/", include("itou.www.dashboard.urls")),
    path("employee_record/", include("itou.www.employee_record_views.urls")),
    path("institutions/", include("itou.www.institutions_views.urls")),
    path("invitations/", include("itou.www.invitations_views.urls")),
    path("prescribers/", include("itou.www.prescribers_views.urls")),
    path("search/", include("itou.www.search.urls")),
    path("siae/", include("itou.www.siaes_views.urls")),
    path("signup/", include("itou.www.signup.urls")),
    path("stats/", include("itou.www.stats.urls")),
    path("versions/", include("itou.www.releases.urls")),
    # Enable Anymailâ€™s status tracking
    # https://anymail.readthedocs.io/en/stable/esps/mailjet/#status-tracking-webhooks
    re_path(r"^webhooks/anymail/", include("anymail.urls")),
    path("welcoming_tour/", include("itou.www.welcoming_tour.urls")),
    # Static pages.
    path("accessibility/", TemplateView.as_view(template_name="static/accessibility.html"), name="accessibility"),
    # Errors pages.
    path("403/", TemplateView.as_view(template_name="403.html"), name="403"),
    path("404/", TemplateView.as_view(template_name="404.html"), name="404"),
    path("500/", TemplateView.as_view(template_name="500.html"), name="500"),
]

if settings.DEBUG and "debug_toolbar" in settings.INSTALLED_APPS:
    import debug_toolbar

    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
    # In the local developpemnt environment, we enable the default callback URLs
    urlpatterns = [path("", include("itou.france_connect.urls"))] + urlpatterns
