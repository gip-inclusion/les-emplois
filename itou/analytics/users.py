from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
