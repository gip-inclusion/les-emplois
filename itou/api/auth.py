from django.contrib.auth.models import AnonymousUser
from rest_framework import authentication

from . import models


class DepartmentTokenAuthentication(authentication.TokenAuthentication):
    model = models.DepartmentToken

    def authenticate_credentials(self, key):
        try:
            api_token = self.model.objects.get(key=key)
            return (AnonymousUser(), api_token)
        except self.model.DoesNotExist:
            # Do not raise AuthenticationFailed to allow other authentication to succeed
            return None
