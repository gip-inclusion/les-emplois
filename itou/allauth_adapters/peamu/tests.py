# Heavily inspired by
# https://github.com/pennersr/django-allauth/blob/master/allauth/socialaccount/providers/google/tests.py
from unittest import mock

from allauth.account.models import EmailAddress, EmailConfirmation
from allauth.account.signals import user_signed_up
from allauth.socialaccount.models import SocialAccount
from allauth.socialaccount.tests import OAuth2TestsMixin
from allauth.tests import MockedResponse, TestCase
from django.core import mail
from django.test import override_settings

from itou.allauth_adapters.peamu.provider import PEAMUProvider
from itou.users import enums as users_enums
from itou.users.models import User


@override_settings(
    API_ESD={"BASE_URL": "https://some.auth.domain", "AUTH_BASE_URL": "https://some-authentication-domain.fr"}
)  # noqa
class PEAMUTests(OAuth2TestsMixin, TestCase):
    provider_id = PEAMUProvider.id

    def get_mocked_response(self, given_name="John", family_name="Doe", email="john.doe@example.com"):
        return MockedResponse(
            200,
            f"""
            {{
                "email": "{email}",
                "gender": "male",
                "given_name": "{given_name}",
                "family_name": "{family_name}",
                "sub": "108204268033311374519"
            }}
            """,
        )

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_data_created_by_peamu_login(self, mock_login_signal):
        test_email = "john.doe@example.com"
        self.login(self.get_mocked_response())
        self.assertEqual(mock_login_signal.call_count, 1)
        email_address = EmailAddress.objects.get(email=test_email)
        account = email_address.user.socialaccount_set.get()
        self.assertEqual(account.extra_data["email"], "john.doe@example.com")
        self.assertEqual(account.extra_data["gender"], "male")
        self.assertEqual(account.extra_data["given_name"], "John")
        self.assertEqual(account.extra_data["family_name"], "Doe")
        # This mysterious 'testac' value actually comes from
        # https://github.com/pennersr/django-allauth/blob/master/allauth/socialaccount/tests.py#L150
        self.assertEqual(account.extra_data["id_token"], "testac")
        self.assertEqual(account.extra_data["sub"], "108204268033311374519")
        self.assertEqual(email_address.user.email, "john.doe@example.com")
        self.assertEqual(email_address.user.first_name, "John")
        self.assertEqual(email_address.user.last_name, "Doe")
        self.assertEqual(email_address.user.username, "john")
        self.assertEqual(email_address.user.is_active, True)
        self.assertEqual(email_address.user.identity_provider, users_enums.IdentityProvider.PE_CONNECT)
        # Note that a PEAMU user is automatically set as a job seeker.
        self.assertEqual(email_address.user.is_job_seeker, True)
        self.assertEqual(email_address.user.is_prescriber, False)
        self.assertEqual(email_address.user.is_siae_staff, False)
        self.assertEqual(email_address.user.is_staff, False)

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_peamu_connection_for_existing_non_peamu_user(self, mock_login_signal):
        email = "user@example.com"
        user = User.objects.create(username="user", is_active=True, email=email, is_job_seeker=True)
        user.set_password("test")
        user.save()
        EmailAddress.objects.create(user=user, email=email, primary=True)
        self.assertFalse(SocialAccount.objects.filter(user=user, provider=PEAMUProvider.id).exists())
        self.client.login(username=user.username, password="test")
        self.assertEqual(mock_login_signal.call_count, 0)

        self.login(self.get_mocked_response(), process="connect")
        # FIXME Seems strange to me
        self.assertEqual(mock_login_signal.call_count, 0)

        self.assertTrue(SocialAccount.objects.filter(user=user, provider=PEAMUProvider.id).exists())
        self.assertEqual(EmailAddress.objects.filter(user=user).count(), 1)
        # User email is not updated.
        self.assertEqual(EmailAddress.objects.filter(email=email).count(), 1)
        self.assertEqual(EmailAddress.objects.filter(email="john.doe@example.com").count(), 0)

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_email_verification_is_skipped_for_peamu_account(self, mock_login_signal):
        test_email = "john.doe@example.com"
        self.login(self.get_mocked_response())
        self.assertEqual(mock_login_signal.call_count, 1)

        email_address = EmailAddress.objects.get(email=test_email)
        self.assertFalse(email_address.verified)
        self.assertFalse(EmailConfirmation.objects.filter(email_address__email=test_email).exists())
        self.assertEqual(len(mail.outbox), 0)

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_username_is_based_on_first_name(self, mock_login_signal):
        first_name = "jessica"
        last_name = "parker"
        email = "john.doe@example.com"
        self.login(self.get_mocked_response(email=email, given_name=first_name, family_name=last_name))
        self.assertEqual(mock_login_signal.call_count, 1)
        user = User.objects.get(email=email)
        self.assertEqual(user.username, "jessica")

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_username_is_based_on_email_if_first_name_is_exotic(self, mock_login_signal):
        first_name = "明"
        last_name = "小"
        email = "john.doe@example.com"
        self.login(self.get_mocked_response(email=email, given_name=first_name, family_name=last_name))
        self.assertEqual(mock_login_signal.call_count, 1)
        user = User.objects.get(email=email)
        self.assertEqual(user.username, "john.doe")

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_user_signed_up_signal(self, mock_login_signal):
        sent_signals = []

        def on_signed_up(sender, request, user, **kwargs):
            sociallogin = kwargs["sociallogin"]
            self.assertEqual(sociallogin.account.provider, PEAMUProvider.id)
            self.assertEqual(sociallogin.account.user, user)
            sent_signals.append(sender)

        user_signed_up.connect(on_signed_up)
        self.login(self.get_mocked_response())
        self.assertEqual(mock_login_signal.call_count, 1)
        self.assertTrue(len(sent_signals) > 0)
