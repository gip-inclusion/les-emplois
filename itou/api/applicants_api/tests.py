from django.urls import reverse
from rest_framework.test import APIClient, APITestCase

from itou.asp.factories import CommuneFactory, CountryFactory
from itou.institutions.factories import InstitutionWithMembershipFactory
from itou.job_applications.factories import JobApplicationFactory
from itou.prescribers.factories import PrescriberOrganizationWithMembershipFactory
from itou.siaes.factories import SiaeFactory
from itou.users.factories import JobSeekerFactory


class ApplicantsAPITest(APITestCase):
    def setUp(self):
        self.client = APIClient()
        self.url = reverse("v1:applicants-list")

    def test_login_as_job_seeker(self):
        user = JobSeekerFactory()
        self.client.force_authenticate(user)

        response = self.client.get(self.url, format="json")
        assert response.status_code == 403

    def test_login_as_prescriber_organisation(self):
        user = PrescriberOrganizationWithMembershipFactory().members.first()
        self.client.force_authenticate(user)

        response = self.client.get(self.url, format="json")
        assert response.status_code == 403

    def test_login_as_institution(self):
        user = InstitutionWithMembershipFactory().members.first()
        self.client.force_authenticate(user)

        response = self.client.get(self.url, format="json")
        assert response.status_code == 403

    def test_api_user_has_unique_siae_membership(self):
        # Connected user must only be member of target SIAE
        user = SiaeFactory(with_membership=True).members.first()
        SiaeFactory(with_membership=True).members.add(user)

        self.client.force_authenticate(user)
        response = self.client.get(self.url, format="json")

        assert response.status_code == 403

    def test_api_user_is_admin(self):
        # Connected user must only be admin of target SIAE
        user = SiaeFactory(with_membership=True).members.first()
        membership = user.siaemembership_set.first()
        membership.is_admin = False
        membership.save()

        self.client.force_authenticate(user)
        response = self.client.get(self.url, format="json")

        assert response.status_code == 403

    def test_login_as_siae(self):
        # Connect with an admin user with member of a sigle SIAE
        user = SiaeFactory(with_membership=True).members.first()
        self.client.force_authenticate(user)

        response = self.client.get(self.url, format="json")
        assert response.status_code == 200
        assert response.json().get("results") == []

    def test_applicant_data(self):
        siae = SiaeFactory(with_membership=True)
        job_seeker = JobApplicationFactory(to_siae=siae).job_seeker
        # Will not refactor ASP factories:
        # - too long,
        # - not the point
        # - scheduled in a future PR
        # => will use some hard-coded values until then
        job_seeker.address_line_1 = "address test"
        job_seeker.address_line_2 = "address 2"
        job_seeker.post_code = "37000"
        job_seeker.city = "TOURS"
        job_seeker.resume_link = "https://myresume.com/me"
        job_seeker.birth_place = CommuneFactory()
        job_seeker.birth_country = CountryFactory()
        job_seeker.save()
        user = siae.members.first()

        self.client.force_authenticate(user)
        response = self.client.get(self.url, format="json")

        assert response.status_code == 200

        [result] = response.json().get("results")

        assert {
            "civilite": job_seeker.title,
            "nom": job_seeker.first_name,
            "prenom": job_seeker.last_name,
            "courriel": job_seeker.email,
            "telephone": job_seeker.phone,
            "adresse": job_seeker.address_line_1,
            "complement_adresse": job_seeker.address_line_2,
            "code_postal": job_seeker.post_code,
            "ville": job_seeker.city,
            "date_naissance": str(job_seeker.birthdate),
            "lieu_naissance": job_seeker.birth_place.name,
            "pays_naissance": job_seeker.birth_country.name,
            "lien_cv": job_seeker.resume_link,
        } == result
