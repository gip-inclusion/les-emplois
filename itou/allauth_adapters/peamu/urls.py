from django.urls import include, re_path

from itou.allauth_adapters.peamu import views as peamu_views
from itou.allauth_adapters.peamu.provider import PEAMUProvider


def default_urlpatterns(provider):
    urlpatterns = [
        re_path(r"^login/$", peamu_views.oauth2_login, name=provider.id + "_login"),
        re_path(r"^login/callback/$", peamu_views.oauth2_callback, name=provider.id + "_callback"),
    ]

    return [
        re_path(r"^profile/$", peamu_views.redirect_to_dashboard_view),
        re_path(r"^" + provider.get_slug() + r"/", include(urlpatterns)),
    ]


urlpatterns = default_urlpatterns(PEAMUProvider)
