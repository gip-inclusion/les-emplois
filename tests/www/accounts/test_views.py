from django.contrib import messages
from django.contrib.auth import get_user
from django.urls import reverse
from pytest_django.asserts import assertContains, assertMessages, assertNotContains

from itou.emails.models import EmailAddress, EmailConfirmation
from tests.users.factories import JobSeekerFactory
from tests.utils.tests import parse_response_to_soup


class TestAccountInactiveView:
    def test_inactive_account(self, client):
        pass


class TestEmailVerificationSentView:
    def test_page_content(self, client, snapshot):
        response = client.get(reverse("accounts:account_email_verification_sent"))
        assert str(parse_response_to_soup(response, selector="#main")) == snapshot


class TestConfirmEmailView:
    def test_confirm_email(self, client):
        # TODO: replace with factory trait with_unverified_email from https://github.com/gip-inclusion/les-emplois/pull/5428
        user = JobSeekerFactory()
        confirm_email_url = EmailConfirmation.create(
            EmailAddress.objects.create(user=user, email=user.email)
        ).get_confirmation_url()

        response = client.get(confirm_email_url)
        assertContains(response, user.email)
        # Soup will raise error if there is no form for POSTing confirmation.
        parse_response_to_soup(response, selector=f"form[action='{confirm_email_url}']")

        # NOTE: Redirect logic is covered by other tests.
        response = client.post(confirm_email_url, {})
        assert get_user(client).is_authenticated
        assertMessages(response, [messages.Message(messages.SUCCESS, f"Vous avez confirm√© {user.email}")])

        email_address = EmailAddress.objects.get(user=user)
        assert email_address.verified
        assert email_address.primary

        confirmation = EmailConfirmation.objects.get(email_address=email_address)
        assert confirmation.used
        assert not confirmation.can_confirm_email()

    def test_logs_out_existing_session(self, client, snapshot):
        """
        A user who is already logged in provides a valid URL. They should be logged out
        """
        client.force_login(JobSeekerFactory())

        new_user = JobSeekerFactory()
        confirm_email_url = EmailConfirmation.create(
            EmailAddress.objects.create(user=new_user, email=new_user.email)
        ).get_confirmation_url()
        client.post(confirm_email_url, {})
        authenticated_user = get_user(client)
        assert authenticated_user.is_authenticated
        assert authenticated_user == new_user

    def test_invalid_url(self, client, snapshot):
        response = client.get(reverse("accounts:account_confirm_email", args=["something-invalid"]))
        assert str(parse_response_to_soup(response, selector="#main")) == snapshot

    # TODO: is this test necessary? If the user has already used the link, then their email will be verified
    # Instead add tests on the model regarding the protections surrounding used.
    def test_used_confirmation(self, client):
        user = JobSeekerFactory()
        confirm_email = EmailConfirmation.create(EmailAddress.objects.create(user=user, email=user.email))
        confirm_email.used = True
        confirm_email.save()

        url = confirm_email.get_confirmation_url()
        response = client.get(url)
        assertNotContains(response, url)
        assert get_user(client).is_authenticated is False

        response = client.post(url)
        assert get_user(client).is_authenticated is False

    def test_confirm_email_verified(self, client):
        # Using a link on an email that's already been verified
        user = JobSeekerFactory()
        email_address = EmailAddress.objects.create(user=user, email=user.email, verified=True)
        confirm_email = EmailConfirmation.create(email_address)

        url = confirm_email.get_confirmation_url()
        response = client.get(url)
        # TODO(calum): Change this test to propose that the user is redirected to ExistingUserLogin.
        assertNotContains(response, url)
        assert get_user(client).is_authenticated is False

        response = client.post(url)
        assert get_user(client).is_authenticated is False
