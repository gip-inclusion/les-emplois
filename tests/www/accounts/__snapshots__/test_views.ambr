# serializer version: 1
# name: TestConfirmEmailView.test_invalid_url
  '''
  <main class="s-main" id="main" role="main">
              <div aria-atomic="true" aria-live="polite" class="toast-container">
  
  </div>
  
              <section class="s-title-02">
                  <div class="s-title-02__container container">
                      <div class="s-title-02__row row">
                          <div class="s-title-02__col col-12">
                              
                              <h1>Confirmer l'adresse e-mail</h1>
                              
      
                                  
  
  
                              
      <div class="alert alert-warning" role="status">
          <p class="mb-0">Ce compte est inactif.</p>
      </div>
  
                              
                          </div>
                      </div>
                  </div>
              </section>
  
              
      <section class="s-section">
          <div class="s-section__container container">
              <div class="row">
                  <div class="col-12 col-lg-8">
                      
  
                          
  
                          <div class="alert alert-danger" role="status">
                              <p class="mb-2">Ce lien de confirmation d'adresse e-mail a expiré ou n'est pas valide.</p>
                              <p class="mb-0">
                                  Veuillez lancer <a href="">une nouvelle demande de confirmation</a>.
                              </p>
                          </div>
  
                      
                  </div>
              </div>
          </div>
      </section>
  
          </main>
  '''
# ---
# name: TestEmailVerificationSentView.test_page_content
  '''
  <main class="s-main" id="main" role="main">
              <div aria-atomic="true" aria-live="polite" class="toast-container">
  
  </div>
  
              <section class="s-title-02">
                  <div class="s-title-02__container container">
                      <div class="s-title-02__row row">
                          <div class="s-title-02__col col-12">
                              
                              
      <h1>Vérifiez votre adresse e-mail</h1>
      <p>
          Nous vous avons envoyé un e-mail avec un lien pour validation. Cliquez sur le lien fourni pour terminer votre inscription.
      </p>
  
                              
                                  
  
  
                              
                              
                          </div>
                      </div>
                  </div>
              </section>
  
              
          </main>
  '''
# ---
# name: TestPasswordResetDoneView.test_already_logged_in
  '''
  <main class="s-main" id="main" role="main">
              <div aria-atomic="true" aria-live="polite" class="toast-container">
  
  </div>
  
              <section class="s-title-02">
                  <div class="s-title-02__container container">
                      <div class="s-title-02__row row">
                          <div class="s-title-02__col col-12">
                              
                              
      <h1>Si un compte existe, vous recevrez un e-mail de réinitialisation</h1>
      
  
                              
                                  
  
  
                              
                              
                          </div>
                      </div>
                  </div>
              </section>
  
              
      <section class="s-section">
          <div class="s-section__container container">
              <div class="row">
                  <div class="col-12 col-lg-8">
                      
                          <p>
      <b>Remarque :</b> vous êtes déjà connecté en tant que <b>jane.doe@test.local</b>.
  </p>
  
                      
                      <p>Il contient les instructions pour réinitialiser votre mot de passe.</p>
                      <p>Si vous le ne recevez pas dans les minutes qui suivent :</p>
                      <ul>
                          <li>vérifiez votre courrier indésirable</li>
                          <li>vérifiez qu'il n'y a pas d'erreur dans votre e-mail</li>
                      </ul>
                  </div>
              </div>
          </div>
      </section>
  
          </main>
  '''
# ---
# name: TestPasswordResetDoneView.test_get
  '''
  <main class="s-main" id="main" role="main">
              <div aria-atomic="true" aria-live="polite" class="toast-container">
  
  </div>
  
              <section class="s-title-02">
                  <div class="s-title-02__container container">
                      <div class="s-title-02__row row">
                          <div class="s-title-02__col col-12">
                              
                              
      <h1>Si un compte existe, vous recevrez un e-mail de réinitialisation</h1>
      
  
                              
                                  
  
  
                              
                              
                          </div>
                      </div>
                  </div>
              </section>
  
              
      <section class="s-section">
          <div class="s-section__container container">
              <div class="row">
                  <div class="col-12 col-lg-8">
                      
                      <p>Il contient les instructions pour réinitialiser votre mot de passe.</p>
                      <p>Si vous le ne recevez pas dans les minutes qui suivent :</p>
                      <ul>
                          <li>vérifiez votre courrier indésirable</li>
                          <li>vérifiez qu'il n'y a pas d'erreur dans votre e-mail</li>
                      </ul>
                  </div>
              </div>
          </div>
      </section>
  
          </main>
  '''
# ---
# name: TestPasswordResetFromKeyDoneView.test_get
  '''
  <main class="s-main" id="main" role="main">
              <div aria-atomic="true" aria-live="polite" class="toast-container">
  
  </div>
  
              <section class="s-title-02">
                  <div class="s-title-02__container container">
                      <div class="s-title-02__row row">
                          <div class="s-title-02__col col-12">
                              
                              
      <h1>Modification de votre mot de passe</h1>
      <p>Votre mot de passe a été modifié !</p>
  
                              
                                  
  
  
                              
                              
                          </div>
                      </div>
                  </div>
              </section>
  
              
          </main>
  '''
# ---
